### PR TITLE
Add an effect that builds&pushes documentation to Github pages

### PR DESCRIPTION
**Pull request by Mr-Andersen**
### Motivation

Provide a simple to use yet configurable Effect

### Maintainer checklist

 - [x] Documentation (function reference docs, setup guide, option reference docs)
 - [ ] Has tests (VM test, free account, and/or test instructions)
 - [ ] Is the corresponding module up to date?

### CODE REVIEW

**Comment by Mr-Andersen** [comment]
>Is the corresponding module up to date?

I don't know what module is the list talking about :grimacing:

File: effects/gh-pages/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# `gh-pages` effect
```
Comment (roberth) [inline effects/gh-pages/README.md:1]: Please read the "top level" review message before diving into the file comments. (The one in the conversation tab)

File: effects/gh-pages/default.nix [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+              mkdir ./gh-pages
+              cd gh-pages
+              git init --initial-branch $branchName
+              git remote add origin $ORIGIN
+            else
+              git clone --branch $branchName --single-branch $ORIGIN gh-pages
+              cd gh-pages
+            fi
+            cp -r --no-preserve=mode $ghPages .
+            git add .
```
Comment (roberth) [inline effects/gh-pages/default.nix:58]: I think you might want to remove existing files, even if not rewriting history.

File: effects/gh-pages/default.nix [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  inherit (pkgs) lib;
+  inherit (pkgs.lib) optionalString;
+in
+
+{ gh-pages
+, branchName ? "gh-pages"
+, condition ? { ref, ... }: lib.elem ref [ "refs/heads/main" "refs/heads/master" ]
+, committer ? {
+    name = "Andrey Vlasov";
+    email = "andreyvlasov+gh-pages-builder@mlabs.city";
```
Comment (roberth) [inline effects/gh-pages/default.nix:16]: This probably shouldn't be the default.
With the modular approach this probably wouldn't even have to be part of the gh pages interface.

**Review by roberth — COMMENTED** [review]
Hi @Mr-Andersen, and thanks for opening this pr!

Recently, I've recently started an effort to modularize git-related effects, with [`flakeUpdate`](https://github.com/hercules-ci/hercules-ci-effects/blob/master/effects/flake-update/effect-fun.nix) as the only, somewhat incomplete example for now. I think it might need one or two extra options to support this use case, but other than that it should help quite a lot to make the implementation (and interface) simpler and more consistent.

You'll find some of my thoughts in the file comments, but I think for now my conclusion is that I need to do some more work to better accommodate effects like this.
That means that this PR isn't immediately actionable, but maybe for now some of the specific comments are helpful for your own copy of this code.

**Comment by roberth** [comment]
> > Is the corresponding module up to date?
> 
> I don't know what module is the list talking about :grimacing:

It's mostly a reminder for me, about flake-parts modules. There's even two types of modules now, so I'll update the template. For now, don't worry about it.

File: effects/gh-pages/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    };
+}
+```
+
+In case you already have effects:
+
+```nix
+herculesCI = herculesEnv: {
+  onPush.myEffect = ...;
+} // mkGhPagesBuilder { inherit (self.packages.${system}) gh-pages; } herculesEnv;
```
[the remark this answers is not in the document]
Comment (Mr-Andersen) [reply effects/gh-pages/README.md:45]: Thank you for noticing bugs like this, I keep forgetting how Nix merges attrsets in different situations

File: effects/gh-pages/default.nix [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  }
+, rewriteHistory ? true
+}:
+{ primaryRepo, ... }:
+{
+  onPush.gh-pages.outputs.effects.default =
```
[the remark this answers is not in the document]
Comment (Mr-Andersen) [reply effects/gh-pages/default.nix:22]: Ok, makes sense, especially considering merging difficulties. I've made `mkGhPagesJob` and `mkGhPagesEffect` functions in my company-private repository

File: effects/gh-pages/default.nix (hunk 3 above)
Comment (Mr-Andersen) [reply effects/gh-pages/default.nix:16]: It'll be enough for internal usage in MLabs :) I would work on this if we were intending to merge this PR

**Comment by Mr-Andersen** [comment]
Thank you for your comments

**Comment by roberth** [comment]
I've implemented the abstractions to make git operations, pushing content and deploying with GitHub Pages easy in
 - https://github.com/hercules-ci/hercules-ci-effects/pull/116

[`deploy-to-github-pages.adoc`](https://github.com/hercules-ci/hercules-ci-effects/blob/gitWriteBranch/docs/modules/ROOT/pages/guide/deploy-to-github-pages.adoc) describes the interface I think you'd want to use.

Would you like to give it a try?

The flakeref is `github:hercules-ci/hercules-ci-effects/gitWriteBranch`

**Pull request closed without merging**